src/julia.h

```diff
@@ -303,36 +300,53 @@ typedef struct _jl_method_t {
     uint8_t pure;
 
 // hidden fields:
-    uint8_t traced;
     // lock for modifications to the method
     jl_mutex_t writelock;
 } jl_method_t;
 
-// This type caches the data for a specType signature specialization of a Method
-struct _jl_method_instance_t {
+// This type is a placeholder to cache data for a specType signature specialization of a Method
+// can can be used as a unique dictionary key representation of a call to a particular Method
+// with a particular set of argument types
+typedef struct _jl_method_instance_t {
     JL_DATA_TYPE
     union {
         jl_value_t *value; // generic accessor
         struct _jl_module_t *module; // this is a toplevel thunk
         jl_method_t *method; // method this is specialized from
-    } def; // context for this lambda definition
+    } def; // pointer back to the context for this code
     jl_value_t *specTypes;  // argument types this was specialized for
-    jl_value_t *rettype; // return type for fptr
-    jl_svec_t *sparam_vals; // static parameter values, indexed by def.method->sig UnionAll tvars
+    jl_svec_t *sparam_vals; // static parameter values, indexed by def.method->sparam_syms
+    jl_value_t *uninferred;
     jl_array_t *backedges;
-    jl_value_t *inferred;  // inferred jl_code_info_t, or jl_nothing, or null
-    jl_value_t *inferred_const; // inferred constant return value, or null
+    struct _jl_code_instance_t *cache;
```

Comment:
Since MethodInstances are already indexed by world age in typemaps, what does it mean to have multiple entries in this cache?

Reply:
The typemap no longer indexes them by world age—that was a failed experiment that led to much confusion (mine and the compilers).